src/Compilers/CSharp/Portable/Symbols/ConstraintsHelper.cs

```diff
@@ -473,9 +473,47 @@ private static bool CheckConstraintsSingleType(TypeSymbol type, CheckConstraints
         // C# does not let you declare a type in which it would be possible for distinct base interfaces
         // to unify under some instantiations.  But such ill-formed classes can come in through
         // metadata and be instantiated in C#.  We check to see if that's happened.
-        private static bool InterfacesAreDistinct(NamedTypeSymbol type, ConsList<Symbol> basesBeingResolved)
+        private static bool HasDuplicateInterfaces(NamedTypeSymbol type, ConsList<Symbol> basesBeingResolved)
         {
-            return !type.InterfacesNoUseSiteDiagnostics(basesBeingResolved).HasDuplicates(TypeSymbol.EqualsIgnoringDynamicComparer);
+            // PERF: avoid instantiating all interfaces here
+            //       Ex: if class implements just IEnumerable<> and IComparable<> it cannot have conflicting implementations
+            var array = type.OriginalDefinition.InterfacesNoUseSiteDiagnostics(basesBeingResolved);
+
+            switch (array.Length)
+            {
+                case 0:
+                case 1:
+                    // less than 2 interfaces
+                    return false;
+
+                case 2:
+                    if (ReferenceEquals(array[0].OriginalDefinition, array[1].OriginalDefinition))
```

Comment:
```(object)array[0].OriginalDefinition == array[1].OriginalDefinition``` ?
Do we care about error scenarios? Error types might have equal but distinct (in terms of reference identity) original definitions.